src/calc/src/expr_manager.rs

```diff
@@ -0,0 +1,540 @@
+            Self::BracketLeft =>  ExprItem::new("(", "(", 4, true, true),
+            Self::BracketRight => ExprItem::new(")", ")", 4, true, true),
+            Self::Comma =>        ExprItem::new(",", ".", 0, true, true),  // FIXME: Maybe we should localize this.
+            Self::Random =>       ExprItem::new("⚄", "random", 3, true, true),
+            Self::Const(name) =>  {
+                // Replace known constants with their characters.
+                match name.as_str() {
+                    "pi" => ExprItem::new("π", "pi()", 0, true, true),
+                    "phi" => ExprItem::new("ϕ", "phi()", 0, true, true),
+                    _ => ExprItem::new(name, format!("{name}()"), 0, true, true)
+                }
+            },
+            Self::Ans => todo!(),
+            _ => return None,
+        })
+    }
+}
+
+/// Holds strings, which are used to convert buttons
+/// to display/evaluate strings.
+#[derive(Debug, Clone)]
+struct ExprItem {
+    disp: String,
+    eval: String,
+
+    // Below properties are used to correctly convert to
+    // evaluate string.
+
+    /// Priority of this function used in `ExprManager::to_postfix()` method.
+    priority: u32,
+    /// Flags if given button is left associative or not.
+    /// If the button is not an operation, then this property
+    /// is ignored.
+    left_asoc: bool,
+    /// Skip convertion to function notation, when converting,
+    /// to evaluation string. Used in `ExprManager::to_eval_str()` method.
+    /// Ignored for non-operation buttons.
+    skip_conv: bool,
+}
+
+impl ExprItem {
+    /// Construct ExprItem from anything convertable to string.
+    pub fn new<T, U>(disp: T, eval: U, priority: u32, left_asoc: bool, skip_conv: bool) -> Self
+    where
+        T: ToString,
+        U: ToString,
+    {
+        Self {
+            disp: disp.to_string(),
+            eval: eval.to_string(),
+            priority,
+            left_asoc,
+            skip_conv,
+        }
+    }
+}
+
+/// Manages mathematical expression, used in our calculation.
+/// We use this, because we intend to have two different
+/// strings for displaying and evaluating.
+#[derive(Debug, Clone)]
+pub struct ExprManager {
+    /// Cursor position in the string.
+    cursor_pos: u32,
+    /// Buttons, which compose the resulting expressoin string.
+    btn_stack: Vec<Btn>,
+    /// Used for invalidating the expression manager (for druid repaint).
+    dirty_flipper: bool,
+}
+
+impl Data for ExprManager {
+    // Dont compare `btn_stack` for performance reasons.
+    fn same(&self, other: &Self) -> bool {
+        self.dirty_flipper == other.dirty_flipper && self.cursor_pos == other.cursor_pos
+    }
+}
+
+impl Default for ExprManager {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ExprManager {
+    /// Create new instance of expression manager with empty
+    /// expression.
+    pub fn new() -> Self {
```

Comment:
can make this as `const` function to evaluate it at compile time

Reply:
Ok